PyDynamic/uncertainty/propagate_filter.py

```diff
@@ -171,95 +181,298 @@ def FIRuncFilter(y, sigma_noise, theta, Utheta=None, shift=0, blow=None, kind="c
+
+    return A, B, C, D
+
+
+def _get_derivative_A(size_A):
+    """
+    build tensor representing dA/dtheta
+    """
+    dA = np.zeros((size_A, size_A, size_A))
+    for k in range(size_A):
+        dA[k, -1, -(k + 1)] = -1
+
+    return dA
+
+
+def _get_corr_unc(b, a, Ux):
+    """
+    Calculate the cumulated correlated noise based on equations (20) of [Link2009]_ .
+    """
+
+    # get impulse response of IIR defined by (b,a)
+    h_theta = np.squeeze(
+        dimpulse((b, a, 1), x0=0.0, t=np.arange(0, len(Ux), step=1))[1][0]
+    )
+
+    # equation (20), note:
+    # - for values r<0 or s<0 the contribution to the sum is zero (because h_theta is zero)
+    # - Ux is the one-sided autocorrelation and assumed to be zero outside its range
+    corr_unc = np.sum(toeplitz(Ux) * np.outer(h_theta, h_theta))
+
+    return corr_unc
+
+
+def get_initial_state(b, a, Uab=None, x0=1.0, U0=1.0, Ux=None):
+    """
+    Calculate the internal state for the IIRuncFilter-function corresponding to stationary
+    non-zero input signal.
+
+    Parameters
+    ----------
+        b: np.ndarray
+            filter numerator coefficients
+        a: np.ndarray
+            filter denominator coefficients
+        Uab: np.ndarray
+            covariance matrix for (a[1:],b)
+        x0: float
+            stationary input value
+        U0: float
+            stationary input uncertainty
+        Ux: np.ndarray, optional
+            single sided autocovariance of stationary (colored/correlated) noise
+            (needed in the `kind="corr"` case of :func:`IIRuncFilter`)
+
+    Returns
+    -------
+    internal_state: dict
+        dictionary of state
+ 
+    """
+
+    # adjust filter coefficients for lengths
+    b, a, Uab = _adjust_filter_coefficients(b, a, Uab)
+
+    # convert into state space representation
+    [A, B, C, D] = _tf2ss(b, a)
+
+    # necessary intermediate variables
+    p = len(A)
+    IminusA = np.eye(p) - A
+    dA = _get_derivative_A(p)
+
+    # stationary internal state
+    # (eye()-A) * zs = B*x0
+    zs = solve(IminusA, B) * x0
+
+    # stationary derivative of internal state
+    # (eye() - A) dzs = dA * zs
+    dzs = solve(IminusA, np.hstack(dA @ zs))
+
+    # stationary uncertainty of internal state
+    Ps = solve_discrete_lyapunov(A, U0 ** 2 * np.outer(B, B))
+
+    if isinstance(Ux, np.ndarray):
+        corr_unc = _get_corr_unc(b, a, Ux)
+    else:
+        corr_unc = 0
+
+    # bring results into the format that is used within IIRuncFilter
+    cache = {
+        "system": (A, B, C, D),
+        "corr_unc": corr_unc,
+        "processed_input": (b, a, Uab),
+    }
+    state = {"z": zs, "dz": dzs, "P": Ps, "cache": cache}
+
+    return state
+
+
+def _adjust_filter_coefficients(b, a, Uab):
+    """
+    Bring b, a to the same length and adjust Uab accordingly
+    """
+    # length difference of both filters
+    d = len(a) - len(b)
+
+    # adjust filter coefficient uncertainties
+    if isinstance(Uab, np.ndarray):
```

Comment:
What else do you expect? We should mention this somewhere here or motivate this check.